src/ORM/Rule/LinkConstraint.php

```diff
@@ -0,0 +1,189 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM\Rule;
+
+use Cake\Datasource\EntityInterface;
+use Cake\ORM\Association;
+use Cake\ORM\Table;
+
+/**
+ * Checks whether links to a given association exist / do not exist.
+ */
+class LinkConstraint
+{
+    /**
+     * Status that requires a link to be present.
+     *
+     * @var string
+     */
+    public const STATUS_LINKED = 'linked';
+
+    /**
+     * Status that requires a link to not be present.
+     *
+     * @var string
+     */
+    public const STATUS_NOT_LINKED = 'notLinked';
+
+    /**
+     * The association that should be checked.
+     *
+     * @var \Cake\ORM\Association|string
+     */
+    protected $_association;
+
+    /**
+     * The link status that is required to be present in order for the check to succeed.
+     *
+     * @var string
+     */
+    protected $_requiredLinkState;
+
+    /**
+     * Constructor.
+     *
+     * @param \Cake\ORM\Association|string $association The alias of the association that should be checked.
+     * @param string $requiredLinkStatus The link status that is required to be present in order for the check to
+     *  succeed.
+     */
+    public function __construct($association, string $requiredLinkStatus)
+    {
+        if (!is_string($association) &&
+            !($association instanceof Association)
+        ) {
+            throw new \InvalidArgumentException(sprintf(
+                'Argument 1 is expected to be of type `\Cake\ORM\Association|string`, `%s` given.',
+                is_object($association) ? get_class($association) : gettype($association)
+            ));
+        }
+
+        if (!in_array($requiredLinkStatus, [static::STATUS_LINKED, static::STATUS_NOT_LINKED], true)) {
+            throw new \InvalidArgumentException(
+                'Argument 2 is expected to match one of the `\Cake\ORM\Rule\LinkConstraint::STATUS_*` constants.'
+            );
+        }
+
+        $this->_association = $association;
+        $this->_requiredLinkState = $requiredLinkStatus;
+    }
+
+    /**
+     * Callable handler.
+     *
+     * Performs the actual link check.
+     *
+     * @param \Cake\Datasource\EntityInterface $entity The entity involved in the operation.
+     * @param array $options Options passed from the rules checker.
+     * @return bool Whether the check was successful.
+     */
+    public function __invoke(EntityInterface $entity, array $options): bool
+    {
+        if (!isset($options['repository']) ||
+            !($options['repository'] instanceof Table)
+        ) {
+            throw new \InvalidArgumentException(
+                'Argument 2 is expected to have a `repository` key that holds an instance of `\Cake\ORM\Table`.'
```

Comment:
You can use Table::class here, it will also remove the unnecessary \ leading slash then :)

Reply:
I could, but that's not the only exception message in this PR that is using a fully qualified name. Is there a specific convention?

Reply:
No, only that we try to use the ::class name to allow the namespace to be pulled by itself.